src/stylus/StylusDeployer.sol

```diff
@@ -0,0 +1,177 @@
+    ///         they should still use a payable caller, since value could get forced into the contract using a methods like self destruct
+    ///         or retryable transactions.
+    ///
+    ///         The caller should do the following before calling this contract:
+    ///         1. Check whether the contract will require activation, and if so what the cost will be.
+    ///            This can be done by spoofing an address with the contract code, then calling ArbWasm.programVersion to compare the
+    ///            the returned result against ArbWasm.stylusVersion. If activation is required ArbWasm.activateProgram can then be called
+    ///            to find the returned dataFee.
+    ///         2. Next this deploy function can be called. The value of the call must be set to the previously ascertained dataFee + initValue
+    ///            If activation is not require, the value of the call should be set to initValue
+    ///
+    ///         Note: Stylus contract caching is not done by the deployer, and will have to be done separately after deployment.
+    ///         See https://docs.arbitrum.io/stylus/how-tos/caching-contracts for more details on caching
+    /// @param bytecode The bytecode of the stylus contract to be deployed
+    /// @param initData Initialisation call data. After deployment the contract will be called with this data
+    ///                 If no initialisation data is provided then the newly deployed contract will not be called.
+    ///                 This means that the fallback function cannot be called from this contract, only named functions can be.
+    /// @param initValue Initialisation value. After deployment, the contract will be called with the init data and this value.
+    ///                  At least as much eth as init value must be provided with this call. Init value is specified here separately
+    ///                  rather than using the msg.value since the msg.value may need to be greater than the init value to accomodate activation data fee.
+    ///                  See the @notice block above for more details.
+    /// @param salt If a non zero salt is provided the contract will be created using CREATE2 instead of CREATE
+    ///             The supplied salt will be hashed with the initData and the initValue so that wherever the address is observed
+    ///             it was initialised with the same variables.
+    /// @return The address of the deployed conract
+    function deploy(
+        bytes calldata bytecode,
+        bytes calldata initData,
+        uint256 initValue,
+        bytes32 salt
+    ) public payable returns (address) {
+        if (salt != 0) {
+            // if a salt was supplied, hash the salt with init value and init data. This guarantees that
+            // anywhere the address of this contract is seen the same init data and value were used.
+            salt = initSalt(salt, initData, initValue);
```

Comment:
I think we should not include `initValue` for consistency with evm CREATE2 factories. Stylus contract init is like constructor initialisation_code for normal evm contract, and CREATE2 does not salt with the deployment value. i.e. 
```
address = keccak256(0xff + sender_address + salt + keccak256(initialisation_code))[12:]
```

Reply:
Updated